### ISSUE TITLE
ColorMapper special colors are not respected for images

### ISSUE THREAD

**Issue by philippjfr** [bokeh/bokeh#5323]
The (Continuous)ColorMapper classes allow defining special colors for the ``nan_color``, ``low_color`` and ``high_color``. However it seems like these are either not respected at all or in some cases even cause errors. Here is a minimal example, I've used named colors here but nothing changes if I use hex colors for example:

```python
import numpy as np
from bokeh.plotting import figure, output_notebook, show
from bokeh.models import LinearColorMapper
from bokeh.palettes import viridis

output_notebook()

p = figure(plot_width=400, plot_height=400, x_range=(0, 10), y_range=(0, 10))

array = np.random.rand(10,10)
array[3:6, 3:6] = np.NaN
cmapper = LinearColorMapper(viridis(256), low=0.25, high=0.75, nan_color='blue', low_color='red', high_color='green')
p.image(image=[array], x=[0], y=[0], dw=[10], dh=[10], color_mapper=cmapper)

show(p)
```

![bokeh_plot 20](https://cloud.githubusercontent.com/assets/1550771/19194416/67848682-8ca5-11e6-9e75-8fb1525fb249.png)

When using an RGB(A) tuple for a color bokeh breaks completely and I have to refresh the page before any plot will render again. The JS console provides the following error message:

```
Uncaught TypeError: t.toLowerCase is not a function
```

Bokeh 0.12.3rc2
OSX 10.11.6
Chrome Version 53.0.2785.143 (64-bit)

**Comment by bryevdv** [comment]
The `file/color_data_mapper.py` is working perfectly on master. I believe this is restricted to some interaction with `Image` which uses colormappers differently. 

The color tuple problem is a separate issue (there have been general probelms with color tuples) There may be an old open issue, if not please open a new one.

**Comment by philippjfr** [comment]
Thanks, should have checked myself if this is just affecting images. Will look into the color tuple issue.

**Comment by bryevdv** [comment]
leaving off the special colors, the color mapper seems to behave as expected. So I would not characterize this as a regression, since the special colors are new. Going to fix after `0.12.3` since no currently existing code should be broken by this.

**Comment by philippjfr** [comment]
I just double checked and using transforms on other glyphs indeed works fine. Color tuples are still broken in that case though.

There are various old issues about this:

https://github.com/bokeh/bokeh/issues/2161
https://github.com/bokeh/bokeh/issues/1381

But not a current one that mentions the error I encountered. Should I open a new issue?

**Comment by bryevdv** [comment]
Yes, let's make a new issue at this point

**Comment by philippjfr** [comment]
Outlined the RGB(A) tuple issue in #5324.

**Comment by bryevdv** [comment]
@philippjfr OK, I think I know where the problem is. The ColorMappers originally had a v_map_screen that was use to map from source (data) to target (colors). This is what Image uses. Then we added computed transforms. Then we wanted to used colomappers in the same way as computed transforms. This was gone by adding on new separate compute method to the colormapper. But it behaves differently than v_map_screen. This raised some flags for me at the time, I should have raised concerns about this, but I fell down in that responsibility.

Long story short: we need to sort out mappers and transforms at a conceptual level first, and put them in a reasonable single, consistent framework. The we need to fix up color mapper in particular. Color mappers should not have have two separate independent interfaces for use.

ping @birdsarah @mattpap @canavandl for thoughts, suggestions, etc.

**Comment by philippjfr** [comment]
Thanks for the detailed explanation, definitely sounds like a good idea to use one consistent interface.
